drivers/base/bootvid/i386/xbox/bootvid.c

```diff
@@ -0,0 +1,461 @@
+    UCHAR BytesPerPixel = 8 * (((NvGetCrtc(Base, 0x19) & 0xE0) << 3) | (NvGetCrtc(Base, 0x13) & 0xFF)) / ScreenWidth;
+
+    if (BytesPerPixel == 4)
+    {
+        ASSERT((NvGetCrtc(Base, 0x28) & 0xF) == BytesPerPixel - 1);
+    }
+    else
+    {
+        ASSERT((NvGetCrtc(Base, 0x28) & 0xF) == BytesPerPixel);
+    }
+
+    return BytesPerPixel;
+}
+
+static VOID
+ApplyPalette(VOID)
+{
+    PULONG Frame = (PULONG)FrameBufferStart;
+    ULONG x, y;
+
+    /* Top panning */
+    for (x = 0; x < PanV * FrameBufferWidth; x++)
+    {
+        *Frame++ = CachedPalette[0];
+    }
+
+    /* Left panning */
+    for (y = 0; y < SCREEN_HEIGHT; y++)
+    {
+        Frame = (PULONG)(FrameBufferStart + (y + PanV) * FrameBufferWidth * BytesPerPixel);
+
+        for (x = 0; x < PanH; x++)
+        {
+            *Frame++ = CachedPalette[0];
+        }
+    }
+
+    /* Screen redraw */
+    PUCHAR Back = BackBuffer;
+    for (y = 0; y < SCREEN_HEIGHT; y++)
+    {
+        Frame = (PULONG)(FrameBufferStart + ((y + PanV) * FrameBufferWidth + PanH) * BytesPerPixel);
+
+        for (x = 0; x < SCREEN_WIDTH; x++)
+        {
+            *Frame++ = CachedPalette[*Back++];
+        }
+    }
+
+    /* Right panning */
+    for (y = 0; y < SCREEN_HEIGHT; y++)
+    {
+        Frame = (PULONG)(FrameBufferStart + ((y + PanV) * FrameBufferWidth + PanH + SCREEN_WIDTH) * BytesPerPixel);
+
+        for (x = 0; x < PanH; x++)
+        {
+            *Frame++ = CachedPalette[0];
+        }
+    }
+
+    /* Bottom panning */
+    Frame = (PULONG)(FrameBufferStart + (PanV + SCREEN_HEIGHT) * FrameBufferWidth * BytesPerPixel);
+    for (x = 0; x < PanV * FrameBufferWidth; x++)
+    {
+        *Frame++ = CachedPalette[0];
+    }
+}
+
+/* PUBLIC FUNCTIONS **********************************************************/
+
+BOOLEAN
+NTAPI
+VidInitialize(
+    _In_ BOOLEAN SetMode)
+{
+    BOOLEAN Result = FALSE;
+
+    /* FIXME: Add platform check */
+
+    /* FIXME: Get access ranges from PnP or IO or HAL */
+    PHYSICAL_ADDRESS PhysControlStart = {.QuadPart = 0xFD000000};
+    PHYSICAL_ADDRESS PhysFrameBufferStart = {.QuadPart = 0xF0000000};
+    ULONG ControlLength = 16 * 1024 * 1024;
+
+    ULONG_PTR ControlStart = (ULONG_PTR)MmMapIoSpace(PhysControlStart, ControlLength, MmNonCached);
+    if (!ControlStart)
+    {
+        DPRINT1("Out of memory!\n");
+        return FALSE;
+    }
+
+    ULONG_PTR FrameBuffer = READ_REGISTER_ULONG((PULONG)(ControlStart + NV2A_CONTROL_FRAMEBUFFER_ADDRESS_OFFSET));
+    FrameBufferWidth = READ_REGISTER_ULONG((PULONG)(ControlStart + NV2A_RAMDAC_FP_HVALID_END)) + 1;
+    FrameBufferHeight = READ_REGISTER_ULONG((PULONG)(ControlStart + NV2A_RAMDAC_FP_VVALID_END)) + 1;
+
+    FrameBuffer &= 0x0FFFFFFF;
+    if (FrameBuffer != 0x3C00000 && FrameBuffer != 0x7C00000)
+    {
+        /* Check framebuffer address (high 4 MB of either 64 or 128 MB RAM) */
+        DPRINT1("Non-standard framebuffer address 0x%p\n", FrameBuffer);
```

Comment:
Not sure it's clever to use any DPRINT functions in bootvid, knowing that one could have enabled debug-on-screen, in which case you would end up somehow with recursive calls.

Reply:
Any idea how to do it properly? Either detect screen debug (not sure, because it's ROS hack) or `#define NDEBUG`?

Reply:
Or use some kind of spinlock 🤔

Reply:
That's going to trigger a deadlock it seems.
bootvid calling that dprint, that calls the debugger in screen mode, that calls bootvid (that then would wait ?)

Reply:
I mean some global variable `BOOLEAN Initialized` with spinlock for read/write access.

Reply:
It's safe since https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/inbv/inbv.c;hb=cd91271796117c0f89c06a6c3f66e6ba4328dd9b#l582